posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,926 @@
+blocking: when a number of transformations are in flight, if one of them is
+much slower than the other ones, [subsequent] transformations will be delayed.
+
+Before we move to another Akka Streams based solution (based on so-called sub-streams),
+it's worth to make a short detour to talk about the facilities that Akka Streams
+provides for logging stuff in a running stream.
+
+== Logging facilities in Akka Streams
+
+Akka Streams has a nice way to log elements (or transformations thereof).
+
+Let's look at how this is done using a simple example:
+
+[source,scala]
+----
+    val piF: Future[BigDecimal] = indexes
+      .log("pi-index", identity)
+      .withAttributes(
+        Attributes.logLevels(
+          onElement = Attributes.LogLevels.Info,
+          onFinish = Attributes.LogLevels.Info,
+          onFailure = Attributes.LogLevels.Error)
+      )
+      .mapAsync(Settings.parallelism)(i => Future(piBBPdeaTermI(i)))
+      .runWith(sumOfTerms)
+----
+
+Assuming you have configured an SLF4J provider (such as logback in the code sample
+repository), logging stuff is done by inserting a `log` combinator which takes two
+arguments: the first is the log name, the second is a function that transforms the
+element in whatever form you're interested in logging. Using the `withAttributes`
+combinator, we can tweak the level at which things are logged. As can be seen from
+the code, we can tweak this level for individual elements as well as for [normal]
+stream completion or stream failure.
+
+In the source code for this example, you will notice that I created an extension
+`logAtInfo` that allows one to add logging in a less verbose manner.
+
+When this extension is applied, the code thus becomes:
+
+[source,scala]
+----
+  val piF: Future[BigDecimal] = indexes
+    .logAtInfo("pi-index")
+    .mapAsync(Settings.parallelism)(piBBPdeaTermIF)
+    .runWith(sumOfTerms)
+----
+
+== Using Substreams to introduce concurrency
+
+An alternative way to speed-up our calculation is to utilise https://doc.akka.io/docs/akka/current/stream/stream-substream.html[Akka Substreams].
+One can consider Substreams as a way to de-multiplex a stream of elements. Substreams
+can be created in different ways, but we'll focus on the `groupBy` combinator.
+`groupBy` takes two arguments, let's start with the second one: this is a function `f`
+that takes an element and which returns a key. The key will determine to which
+Substream the element will be sent. The idea is that `f` returns a finite number
+of unique key values. The first parameter `maxSubstreams` is the number of distinct
+key values generated by the function. Note that if `f` generates more distinct key
+values than `maxSubstreams`, the stream will fail. The following code segment shows
+the splitting of our main stream into Substreams:
+
+[source,scala]
+----
+  val piF = indexes
+    .groupBy(Settings.parallelism, _ % Settings.parallelism)
```

Comment:
Perhaps using named arguments (and some type annotations) would make this easier to follow and understand?